op-batcher: extract state pruning, block fetching and progress checking into a single pure function

geoknee opened the pull request:
This approach enables the majority of the complexity in these jobs to be encapsulated, deduplicated, and easily unit tested.

Supercedes #12981.

geoknee committed:
remove lastStoredBlock and lastL1Tip from BatchSubmitter state

We can use the channelManager's state to infer lastStoredBlock. And lastL1Tip is actually unused.

geoknee committed:
change log line wording

geoknee committed:
fix typo

geoknee committed:
remove unecessary method

geoknee committed:
WIP first pass at computeSyncActions

geoknee committed:
computeSyncAction takes a ChannelStatuser interface

also report fully inclusive range of blocks to load

geoknee committed:
add happy path test case

geoknee committed:
Merge branch 'gk/new-batcher-fix' into gk/compute-sync-actions

geoknee committed:
clearState is a pointer

we can use nil value to signal no state clearing should be performed

geoknee committed:
add more test cases

geoknee committed:
add another test case

geoknee committed:
computeSyncActions only takes prevCurrentL1, not prevSyncStatus

geoknee committed:
add batcher restart case

geoknee committed:
safe chain reorg case

geoknee committed:
failed to make progress case

geoknee committed:
simplify log messages, print entire struct

geoknee committed:
add godoc

geoknee committed:
wire up computeSyncActions

geoknee committed:
cache prevCurrentL1 on BatchSubmitter

geoknee committed:
document stages

geoknee committed:
fix loadBlocksIntoState range interpretation

geoknee committed:
pass syncStatus, not pointer to syncStatus and add test case for no progress

geoknee committed:
check unsafe status before trying to get more blocks

geoknee committed:
do not panic on invalid block ranges

return an error instead. This error is ultimated swallowed, matching existing behaviour.

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/compute-sync-actions

geoknee commented on op-batcher/batcher/driver_test.go:
```diff
@@ -116,3 +119,179 @@ func TestBatchSubmitter_SafeL1Origin_FailsToResolveRollupClient(t *testing.T) {
+			expected:      SyncActions{waitForNodeSync: true},
+			expectedLogs:  []string{"sequencer currentL1 reversed"},
+		},
+		{name: "L1",
```
This is not a very descriptive name, do I mean "L1 reorg"?

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +919,136 @@ func logFields(xs ...any) (fs []any) {
+	MaxInclusionBlock() uint64
+}
+
+type SyncActions struct {
+	clearState      *eth.BlockID
+	blocksToPrune   int
+	channelsToPrune int
+	waitForNodeSync bool
+	blocksToLoad    [2]uint64 // the range [start,end] that should be loaded into the local state.
+	// NOTE this range is inclusive on both ends, which is a change to previous behaviour.
+}
```
From chat with @sebastianst today, perhaps we could change the API so that `waitForNodeSync` and `clearState` could be extracted into two different error values returned This might make the data structure match the different actions better, and make it impossible to do incorrect things. 

Or, going in the other direction, should we _always_ wait for node sync before taking any action?

geoknee replied:
It might be better for the error to be "needs node sync" or "sequencer out of sync". The caller can then decide how to deal with that error, e.g. either:
1. Wait another ticker
2. or wait _now_ for the sequencer to catch up

geoknee replied:
I don't know if `clearState` should be an error because we actually know which blocks to load in this case and can (I think) safely load those blocks after clearing the state. So we would be returning non-empty `blocksToLoad` _and_ an error, which doesn't seem very idiomatic?

geoknee replied:
Made the "needs sync" case an error here [47578b2](https://github.com/ethereum-optimism/optimism/pull/13060/commits/47578b23642403a90702663723751495854e4b58)

geoknee replied:
Also, `clearState` does need to specify a value (a pointer in fact) which isn't just a boolean, so it seems less natural to make it an error.

sebastianst replied:
I think it makes sense to not have `clearState` be an error, but instead be an optional _action_ 👍🏻

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +919,136 @@ func logFields(xs ...any) (fs []any) {
+		// The sequencer may have derived the safe chain
+		// from channels sent by a previous batcher instance.
+		s := SyncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
```
What's the correct value for this? Can we update the readme to describe how this mechanism works (e.g. to try and keep a steady pulse of batcher transactions). 

Is `waitNodeSyncAndClearState()` now dead code? Can we remove it?

geoknee replied:
`waitNodeSyncAndClearState()` is still used for when we try to get blocks from the sequencer. That can't be encapsulated into the new pure function, but can still indicate a reorg.

geoknee replied:
Updated the readme here [ba62afe](https://github.com/ethereum-optimism/optimism/pull/13060/commits/ba62afe93b5e0687cea5cdd661aecae8563479b2)

geoknee commented on op-batcher/batcher/driver_test.go:
```diff
@@ -116,3 +119,179 @@ func TestBatchSubmitter_SafeL1Origin_FailsToResolveRollupClient(t *testing.T) {
+			channels:      []ChannelStatuser{channel103},
+			expected: SyncActions{
+				waitForNodeSync: true,
+				clearState:      &eth.BlockID{},
```
None of these test cases returns a non empty value for `clearState`. Does that mean we need more coverage? Are the test assertions definitely working correctly?

geoknee replied:
Why do so many of theses cases return an empty / zero valued but non nil `*eth.BlockID{}`? When would we want to do that? (It would trigger the batcher to clear its state entirely, and then submit a transaction as soon as possible afterward because it would mark the channel full immediately).

geoknee replied:
It almost never makes sense to use an empty value for this.

geoknee replied:
The problem is that the test is passing incomplete mocked data. We should fix that, and the assertions.

geoknee replied:
[b8a43b5](https://github.com/ethereum-optimism/optimism/pull/13060/commits/b8a43b50b85410d6c216169cfbdaa81dff4511af)

geoknee commented on op-batcher/batcher/channel_manager_test.go:
```diff
@@ -462,149 +462,6 @@ func TestChannelManager_handleChannelInvalidated(t *testing.T) {
 
-func TestChannelManager_PruneBlocks(t *testing.T) {
-	l := testlog.Logger(t, log.LevelDebug)
-	cfg := channelManagerTestConfig(100, derive.SingularBatchType)
```
We can add back some basic tests for this.

geoknee replied:
[6d41fab](https://github.com/ethereum-optimism/optimism/pull/13060/commits/6d41fab1146581ba59e258e5269c6dc5348b78c9)

geoknee commented on op-batcher/batcher/channel_manager_test.go:
```diff
@@ -462,149 +462,6 @@ func TestChannelManager_handleChannelInvalidated(t *testing.T) {
-
-}
-
-func TestChannelManager_PruneChannels(t *testing.T) {
```
Let's add back some basic tests for this.

geoknee replied:
[6d41fab](https://github.com/ethereum-optimism/optimism/pull/13060/commits/6d41fab1146581ba59e258e5269c6dc5348b78c9)

geoknee committed:
test: add assetions and mock data about blockID passed to clearState

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +919,136 @@ func logFields(xs ...any) (fs []any) {
+// state of the batcher (blocks and channels), the new sync status, and the previous current L1 block. The actions are returned
+// in a struct specifying the number of blocks to prune, the number of channels to prune, whether to wait for node sync, the block
+// range to load into the local state, and whether to clear the state entirely.
+func computeSyncActions[T ChannelStatuser](newSyncStatus eth.SyncStatus, prevCurrentL1 eth.L1BlockRef, blocks queue.Queue[*types.Block], channels []T, l log.Logger) SyncActions {
```
Wdyt about renaming this function to something like `syncStep`, because of its similarities to the derivation pipeline step function?

geoknee replied:
I'm not sure it makes a lot of sense here, since we don't have "steps" and multiple pipeline stages? Open to renaming in general though if we don't like `computeSyncActions`.

geoknee committed:
add readme section on max channel duration

geoknee committed:
add back unit tests for pruning methods

geoknee committed:
fix pruneBlocks behaviour when blockCursor pointed at block which is now pruned

geoknee committed:
rename waitForNodeSync to sequencerOutOfSync

geoknee committed:
Introduce SeqOutOfSyncError

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
 	return fs
 }
+
+type ChannelStatuser interface {
```
Because two of the methods of this interface are private, I suggest we make this type itself private. A different package couldn't implement it anyways.

geoknee replied:
[ChannelStatuser -> channelStatuser](https://github.com/ethereum-optimism/optimism/pull/13060/commits/c891505c28f77adce52d369a8314641530f4e686)

sebastianst commented on packages/contracts-bedrock/lib/lib-keccak:
was this update intended?

geoknee replied:
No, [undo changes to submodule](https://github.com/ethereum-optimism/optimism/pull/13060/commits/fa5fd37861f2a330bc9807fb3ad8cf6e6245ab5d)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
+	MaxInclusionBlock() uint64
+}
+
+var SeqOutOfSyncError error = errors.New("sequencer out of sync")
```
In idiomatic Go, only error _types_ are `FooError`s. This should be
```suggestion
var ErrSeqOutOfSync error = errors.New("sequencer out of sync")
```

geoknee replied:
[SeqOutOfSyncError -> ErrSeqOutOfSync](https://github.com/ethereum-optimism/optimism/pull/13060/commits/1ed9fff79abbc9347c34c794f1d64951c27a9b59)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -241,28 +243,15 @@ func (l *BatchSubmitter) StopBatchSubmitting(ctx context.Context) error {
-		return errors.New("start number is >= end number")
+// loadBlocksIntoState loads the blocks between start and end (inclusive).
+// If there is a reorg, it will return an error.
+func (l *BatchSubmitter) loadBlocksIntoState(start, end uint64, ctx context.Context) error {
```
Let's move `ctx` as the first parameter, which is more idiomatic. Good opportunity to fix.

geoknee replied:
[move ctx to first position in fn signature](https://github.com/ethereum-optimism/optimism/pull/13060/commits/e5dee3409e7cb206df44252909578bade5c4dc6b)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -464,20 +425,34 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
-				l.waitNodeSyncAndClearState()
+			// Decide appropriate actions
+			syncActions, err := computeSyncActions(*syncStatus, l.prevCurrentL1, l.state.blocks, l.state.channelQueue, l.Log)
+			l.prevCurrentL1 = syncStatus.CurrentL1
```
Shouldn't we only update this if we don't get a seq-out-of-sync error?

geoknee replied:
I think you are right [do not update cached prevCurrentL1 value if there is an ErrSeqOutOfSync](https://github.com/ethereum-optimism/optimism/pull/13060/commits/eca021c6b47b8634ad3771699b2f5f354d0e20f9)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -464,20 +425,34 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
+				// If the sequencer is out of sync
+				// do nothing and wait to see if it has
+				// got in sync on the next tick.
+				continue
```
Let's still add a warn-log here.

geoknee replied:
[Always warn log when computeSyncActions returns an error](https://github.com/ethereum-optimism/optimism/pull/13060/commits/b3337fbc5334e89acedf24c639b43cdcdb382130) but then also [computeSyncActions returns a bool, not an error](https://github.com/ethereum-optimism/optimism/pull/13060/commits/8a0824c3bc7d99c6c08fc34025f10d6a2e8b82e5)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
+
+var SeqOutOfSyncError error = errors.New("sequencer out of sync")
+
+type SyncActions struct {
```
I'd make this a private type, it's only used package-internally and is returned by an internal function.

geoknee replied:
[SyncActions -> syncActions](https://github.com/ethereum-optimism/optimism/pull/13060/commits/337a4ec2e5d0431d2263c373b0ef02d945bfcbdc)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -464,20 +425,34 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
-				l.waitNodeSyncAndClearState()
-				continue
+			// Manage existing state / garbage collection
+			if syncActions.clearState != nil {
+				l.state.Clear(*syncActions.clearState)
+			} else {
+				l.state.pruneSafeBlocks(syncActions.blocksToPrune)
+				l.state.pruneChannels(syncActions.channelsToPrune)
+			}
```
So state clearing and pruning are mutually exclusive?

geoknee replied:
Yes, clearing is essentially doing (at least) a full prune anyway so there's not point doing both.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
+// state of the batcher (blocks and channels), the new sync status, and the previous current L1 block. The actions are returned
+// in a struct specifying the number of blocks to prune, the number of channels to prune, whether to wait for node sync, the block
+// range to load into the local state, and whether to clear the state entirely. Returns an error if the sequencer is out of sync.
+func computeSyncActions[T ChannelStatuser](newSyncStatus eth.SyncStatus, prevCurrentL1 eth.L1BlockRef, blocks queue.Queue[*types.Block], channels []T, l log.Logger) (SyncActions, error) {
```
it's annoying that Go forces us to do tricks like making a generic function just so that we don't need to convert between struct and interface slices 😅

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
+
+	if !hasBlocks {
+		s := SyncActions{
+			blocksToLoad: [2]uint64{newSyncStatus.SafeL2.Number + 1, newSyncStatus.UnsafeL2.Number},
```
`newSyncStatus.SafeL2.Number + 1` is used a lot in this function, maybe we save it to a separate variable like `nextSafeL2 = newSyncStatus.SafeL2.Number + 1`? May help reading the function.

geoknee replied:
[define local variables to aid readability](https://github.com/ethereum-optimism/optimism/pull/13060/commits/4a2cf6903a85146ec9abf82a2b56910da58e4cf5)

sebastianst reviewed (commented):
forgot to submit this review...

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
+		return SyncActions{}, SeqOutOfSyncError
+	}
+
+	oldestBlock, hasBlocks := blocks.Peek()
```
remind me, the `blocks` can only be empty if the channels are empty as well? so all blocks in pending channels are still in the `blocks`?

geoknee replied:
Yes, blocks are now _copied_ into channels, not _moved_. And they are only removed by the new "pruning" functions, which should clean up blocks and channels together in the happy path. In the unhappier paths where we are clearing state, then they will both be completely reset anyway.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
+	}
+
+	if oldestBlock.NumberU64() > newSyncStatus.SafeL2.Number+1 {
+		s := SyncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
+			blocksToLoad: [2]uint64{newSyncStatus.SafeL2.Number + 1, newSyncStatus.UnsafeL2.Number},
+		}
```
It looks like this exact `SyncAction` is the same in the next 4 or so clear-state cases. So maybe we save it to a `clearSyncAction` and then we can just return the same in the next couple clear-state cases.

geoknee replied:
[organise computeSyncActions and introduce startAfresh syncAction](https://github.com/ethereum-optimism/optimism/pull/13060/commits/8892aa5b4e06fc2ab74859a1eb49be7986d3a9df)

geoknee committed:
move SyncActions code to a separate file

geoknee committed:
ChannelStatuser -> channelStatuser

geoknee committed:
SeqOutOfSyncError -> ErrSeqOutOfSync

geoknee committed:
move ctx to first position in fn signature

geoknee committed:
do not update cached prevCurrentL1 value if there is an ErrSeqOutOfSync

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -114,7 +115,8 @@ type BatchSubmitter struct {
 	txpoolBlockedBlob bool
 
-	state *channelManager
+	state         *channelManager
+	prevCurrentL1 eth.L1BlockRef // cached CurrentL1 from the last syncStatus
```
Wondering about pusing this `prevCurrentL1` variable inside `state`, so that `state` encapsulates all of the actual state. This would mean that it gets cleared when we decide to "clear state" however, which might not be what we want.

sebastianst replied:
I think it's better to keep it the way it currently is, because the `prevCurrentL1` matters in the context of `SyncStatus` handling, which happens at the level of the batcher main loop, not inside the channel manager. We should maybe instead rename the `state` variable just to `channelMgr` or so, because `state` is misleading 😅

geoknee committed:
Always warn log when computeSyncActions returns an error

geoknee committed:
move sync actions test to separate file

geoknee committed:
computeSyncActions returns a bool, not an error

There is only ever one kind of error returned

geoknee committed:
SyncActions -> syncActions

geoknee committed:
define local variables to aid readability

geoknee committed:
organise computeSyncActions and introduce startAfresh syncAction

Add comments explaining logical flow: the checks get increasingly deep and we return early where possible.

geoknee committed:
undo changes to submodule

geoknee committed:
move test utils to sync_actions_test.go file

geoknee committed:
ensure pruneChannels clears currentChannel when appropriate

geoknee committed:
fix submodule"

geoknee committed:
don't try to get number of block if none exists

geoknee committed:
improve log

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -464,20 +424,34 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
+				// If the sequencer is out of sync
+				// do nothing and wait to see if it has
+				// got in sync on the next tick.
+				l.Log.Warn("error computing sync actions, sequencer is out of sync")
```
It's not really an error, since it's expected behavior during a node restart. Maybe something like
```suggestion
				l.Log.Warn("Sequencer is out of sync, retrying next tick.")
```

sebastianst commented on op-batcher/batcher/sync_actions.go:
```diff
@@ -0,0 +1,144 @@
+	clearState      *eth.BlockID
+	blocksToPrune   int
+	channelsToPrune int
+	blocksToLoad    [2]uint64 // the range [start,end] that should be loaded into the local state.
```
I'm not a fan of using an array for a range. I'd find it clearer to just use either two fields like `loadBlocksStart` and `loadBlocksEnd` or make this a `*struct{start, end uint64}`. Wdyt?

Sorry for just remarking this now.

geoknee replied:
[f588a00](https://github.com/ethereum-optimism/optimism/pull/13060/commits/f588a009c75168ef5819fd8868ebfb711db4b71c)

sebastianst commented on op-batcher/batcher/sync_actions.go:
```diff
@@ -0,0 +1,144 @@
+	oldestBlockInStateNum := oldestBlockInState.NumberU64()
+
+	if oldestUnsafeBlockNum < oldestBlockInStateNum {
+		s := syncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
+			blocksToLoad: [2]uint64{oldestUnsafeBlockNum, youngestUnsafeBlockNum},
+		}
```
you could also use `startAfresh` here.

geoknee replied:
[c534f92](https://github.com/ethereum-optimism/optimism/pull/13060/commits/c534f92540a70a39cbda91d1d475ef1f0fe7df25)

sebastianst commented on op-batcher/batcher/sync_actions.go:
```diff
@@ -0,0 +1,144 @@
+		}
+	}
+
+	// PART 5: happy path
+	numChannelsToPrune := 0
+	for _, ch := range channels {
+		if ch.LatestL2().Number > newSyncStatus.SafeL2.Number {
+			break
+		}
+		numChannelsToPrune++
+	}
```
If I see correctly, all test cases only involve a single channel. I think it would be good to have a test case that has multiple channels where only a subset is determined to be pruned, to test correctness of this loop.

geoknee replied:
[4df01e1](https://github.com/ethereum-optimism/optimism/pull/13060/commits/4df01e152757087e710207a1ca561a25d7bff637)

sebastianst reviewed (commented):
looking good, just minor comments

geoknee committed:
Update op-batcher/batcher/driver.go

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

geoknee committed:
use struct for block range, not array

geoknee committed:
use startAfresh in one more place

geoknee committed:
add test case for multiple channels

also set HeadL1 to more realistic values (generally ahead of currentL1 due to nonzero confirmation depth)

sebastianst commented on op-batcher/batcher/sync_actions.go:
```diff
@@ -16,17 +16,18 @@ type channelStatuser interface {
 func (s syncActions) String() string {
 	return fmt.Sprintf(
-		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, clearState: %v, blocksToLoad: [%d, %d]}", s.blocksToPrune, s.channelsToPrune, s.clearState, s.blocksToLoad[0], s.blocksToLoad[1])
+		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, clearState: %v, blocksToLoad: [%d, %d]}", s.blocksToPrune, s.channelsToPrune, s.clearState, s.blocksToLoad.start, s.blocksToLoad.end)
```
This will panic if the range is `nil`.

geoknee replied:
[b1f5fe7](https://github.com/ethereum-optimism/optimism/pull/13060/commits/b1f5fe7d6751871deca5da95c2f5d0d691689cca)

geoknee committed:
print value of *struct  in Stringer

geoknee commented:
Going by the codecov report: we should add a test case where there are no blocks in the state. That should give us 100% coverage of the new pure function.

geoknee committed:
add test case when there are no blocks in state

sebastianst commented on op-batcher/batcher/sync_actions.go:
```diff
@@ -27,7 +27,7 @@ type syncActions struct {
 func (s syncActions) String() string {
 	return fmt.Sprintf(
-		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, clearState: %v, blocksToLoad: [%d, %d]}", s.blocksToPrune, s.channelsToPrune, s.clearState, s.blocksToLoad.start, s.blocksToLoad.end)
+		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, clearState: %v, blocksToLoad: [%v]}", s.blocksToPrune, s.channelsToPrune, s.clearState, s.blocksToLoad)
```
```suggestion
		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, clearState: %v, blocksToLoad: %v}", s.blocksToPrune, s.channelsToPrune, s.clearState, s.blocksToLoad)
```
otherwise you'll have `[...]` around the struct print

geoknee committed:
Update op-batcher/batcher/sync_actions.go

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

geoknee committed:
tighten up log messages and test descriptions

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -464,20 +424,34 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
+				// Get fresh unsafe blocks
+				if err := l.loadBlocksIntoState(l.shutdownCtx, syncActions.blocksToLoad.start, syncActions.blocksToLoad.end); errors.Is(err, ErrReorg) {
+					l.Log.Warn("error loading blocks, clearing state and waiting for node sync", "err", err)
+					l.waitNodeSyncAndClearState()
```
I am wondering if we want to keep this alternative way of "waiting for sync" and "clearing state:. 

Above, we are now waiting for sync purely in the sense that the sync status needs to be non empty and the currentl1 needs to not have reversed. Here, we wait for transactions to be ingested by the sequencer.

Above, we clear the `channelManager` state with the safel2.l1origin that we have from the sync status already in scope. Here, we go ahead and query for that again before clearing in the same way. 

Feels like an opportunity to harmonize, even delete some code?

sebastianst replied:
It's a good observation, definitely looks like something could be harmonized here. 

What makes this non-trivial is that we cannot detect the reorg case with 100% certainty earlier inside `computeSyncActions`. There might be a reorg between the time the block range is determined and then when we're actually fetching the blocks. So we have to check for reorgs at the time when we're actually processing the fetches blocks by calling `channelManager.AddL2Block` and see whether it returns an `ErrReorg`.

Maybe there's an opportunity to let the next `computeSyncActions` step handle the reorg in the next loop iteration, instead of handling it right there. But certainly needs some thought. I suggest we reserve this for a follow up and get this in as it is now.

geoknee replied:
https://github.com/ethereum-optimism/optimism/issues/13279

sebastianst replied:
tracking: https://github.com/ethereum-optimism/optimism/issues/13279
